:language: r
:source-highlighter: pygments
:pygments-linenums-mode: table
:toc2:
:numbered:
:experimental:
:data-uri:
:icons: font

= Practical course in pan-genomics, transcriptomics and functional genomics

Martin Mascher, IPK Gatersleben, November 2025
mascher@ipk-gatersleben.de

++++
<link rel="stylesheet"  href="http://cdnjs.cloudflare.com/ajax/libs/font-awesome/3.1.0/css/font-awesome.min.css">
++++

//== Software installation

//IMPORTANT: The Virtual Machine has already been set up with all necessary software installed for the course at CIHEAM Zaragoza in November 2025. 

//. All the commands were tested on an Ubuntu virtual machine (VM) with standard software configuration (Ubuntu 20.04.1 LTS). The VM has 8 GB of main memory and runs on an Intel(R) Xeon(R) Gold 5218 CPU (2.30GHz, 2 physical cores), and has access to 500 GB of hard disk space. Fewer than 8 GB of RAM would prevent one programs we will use (minimap2) to run properly, but most steps should work with 4 GB or even 2 GB RAM. A total of 70 GB of disk space are needed to store the RNA-seq dataset we will analyze.

//. Information on how to setup an Ubuntu VM were given in the 
//https://github.com/eead-csic-compbio/scripting_linux_shell[course] of Bruno Contreras and Carlos Cantalapiedra. See their https://github.com/eead-csic-compbio/scripting_linux_shell/blob/master/session0.md[session 0].

//. Install the required tools with the http://manpages.ubuntu.com/manpages/bionic/man8/apt.8.html[apt] package manager.
//You will be prompted to enter your user password.
//+
//[source,sh]
//----
//sudo apt install samtools cufflinks kallisto minimap2 python git r-base ncbi-blast+ libssl-dev curl libcurl4-openssl-dev pandoc emboss mafft snp-sites gmap
//----
//
//.. http://samtools.github.io[samtools], https://github.com/lh3/minimap2[minimap2], https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download[ncbi-blast+], http://emboss.sourceforge.net[emboss], https://mafft.cbrc.jp/alignment/software/[mafft], http://sanger-pathogens.github.io/snp-sites/[snp-sites], http://research-pub.gene.com/gmap/[gmap] and http://cole-trapnell-lab.github.io/cufflinks/[cufflinks] will be used in the analysis of two rice genomes.
//
//.. https://pachterlab.github.io/kallisto/[kallisto] will be used to analyse RNAseq data of wheat.


== Comparing two rice genome assemblies

=== Preparing the data

. Enter the folder with the data for this course.
+
[source,sh]
----
cd /home/iamz/materials_Mascher
----

. Have a look at the IR64 assembly. It is described in this https://www.g3journal.org/content/10/5/1495[paper]. Have a look at the file.
+
[source,sh]
----
ls IR64.fasta
less IR64.fasta
----

. Have a look at the IR64 assembly. It is described in this https://thericejournal.springeropen.com/articles/10.1186/1939-8433-6-4[paper]. Have a look at the file.
+
[source,sh]
----
ls Nipponbare.fasta
less Nipponbare.fasta
----

=== Chromosome-wise alignment

. Create the working directory and change into it.
+
[source,sh]
----
mkdir -p ~/rice_assemblies
cd ~/rice_assemblies
----

. Create symbolic links to the FASTA files of the assemblies in the downloads folder.
+
[source,sh]
----
ln -s /home/iamz/materials_Mascher/IR64.fasta
ln -s /home/iamz/materials_Mascher/Nipponbare.fasta
----

. Create index files that record the chromosome lengths in each assembly.
+
[source,sh]
----
samtools faidx IR64.fasta
samtools faidx Nipponbare.fasta
----

. Align the two assemblies with https://github.com/lh3/minimap2[minimap2].
+
[source,sh]
----
minimap2 -I500M -x asm5 Nipponbare.fasta IR64.fasta > Nipponbare_IR64.paf 2> Nipponbare_IR64.paf.err  #<1><2><3>
----
<1> The parameter preset `-x asm5` is suited to align assemblies from the same species. 
<2> The value of `-I` is reduced to 500 MB from the default 4 GB to reduce the memory footprint.
<3> `>` and `2>` are used to direct the standard output (stdout) and standard error (stderr).

. Send the `minimap2` process to the background with Ctrl-Z. Resume it with `bg`. Check with `top` that it's running.

. In the meantime, take a look the definition of the PAF format at the end of the minimap2 manpage.
+
[source,sh]
----
man minimap2
----

. Once the `minimap2` is done, have a look at output.
+
[source,sh]
----
cat Nipponbare_IR64.paf
head Nipponbare_IR64.paf
less Nipponbare_IR64.paf
less -S Nipponbare_IR64.paf
column -t Nipponbare_IR64.paf | less -S
----

//. Copy the PAF file to the Windows desktop and open them.
//+
//[source,sh]
//----
//cp Nipponbare_IR64.paf /media/Desktop/
//----

. Load the PAF file into the https://tom-poorten.shinyapps.io/dotplotly_shiny/[Dotplotly] app. 

=== Comparing a single gene in two assemblies

. Create a symbolic link to the FASTA file with the sequence of the PLASTOCHRON1 gene (https://dx.doi.org/10.1073%2Fpnas.2636936100[PLASTOCHRON1]). Take a look at the sequence.
+
[source,sh]
----
ln -s /home/iamz/materials_Mascher/pla1.fasta 
less pla1.fasta
----

. Create BLAST database for the two genome assemblies.
+
[source,sh]
----
makeblastdb -dbtype nucl -in Nipponbare.fasta
makeblastdb -dbtype nucl -in IR64.fasta
----

. Run the BLAST alignment and output to http://www.metagenomics.wiki/tools/blast/blastn-output-format-6[tabular format].
+
[source,sh]
----
blastn -db Nipponbare.fasta -query pla1.fasta -outfmt 6 > pla1_Nipponbare.txt #<1>
blastn -db IR64.fasta -query pla1.fasta -outfmt 6 > pla1_IR64.txt
----
<1> `-outfmt 6` means http://www.metagenomics.wiki/tools/blast/blastn-output-format-6[tabular] output.

. Extract the aligned sequence of the first exon with `samtools faidx` [http://www.htslib.org/doc/samtools-faidx.html[man page]].
+
[source,sh]
----
samtools faidx IR64.fasta chr10:10413299-10414334 > pla1_IR64.fasta
samtools faidx Nipponbare.fasta Chr10:13659508-13660543 > pla1_Nipponbare.fasta
----

. Extract the sequence of all BLAST hits.
+
[source,sh]
----
cat pla1_IR64.txt | awk '$9 < $10 {print $2":"$9"-"$10} $10 < $9 {print $2":"$10"-"$9}' | sort | xargs samtools faidx  IR64.fasta  > pla1_IR64_all_hits.fasta 
----

. Align the two sequence using MAFFT and find SNPs between them with SNP-sites:
+
[source,sh]
----
cat pla1_IR64.fasta pla1_Nipponbare.fasta | mafft - > pla1_mafft.fasta
snp-sites -v pla1_mafft.fasta
snp-sites pla1_mafft.fasta
----

. Now we use https://academic.oup.com/bioinformatics/article/21/9/1859/409207[GMAP] for spliced alignment 
to extract and compare alignments of of the full transcript, not only the first exon.

. Build the GMAP indices for both genomes.
+
[source,sh]
----
gmap_build Nipponbare.fasta -D . -d Nipponbare_db > Nipponbare_build.out 2> Nipponbare_build.err & #<1>
gmap_build IR64.fasta -D . -d IR64_db > IR64_build.out 2> IR64_build.err &
----
<1> The `&` at the end of the line sends the process immediately to the background.

. Align the PLA1 sequence to both genomes.
+
[source,sh]
----
gmap -f 2 -D . -d Nipponbare_db pla1.fasta > pla1_Nipponbare_gmap.gff #<1>
gmap -f 2 -D . -d IR64_db pla1.fasta > pla1_IR64_gmap.gff
----
<1> `-f 2` generate GFF output. A description of the GFF format can be found https://www.ensembl.org/info/website/upload/gff.html[here].

. Extract the matched sequence with https://github.com/gpertea/gffread[gffread] (part of http://cole-trapnell-lab.github.io/cufflinks/[Cufflinks]).
+
[source,sh]
----
cat pla1_Nipponbare_gmap.gff | gffread -g Nipponbare.fasta -w pla1_Nipponbare_gmap_mRNA.fasta
cat pla1_IR64_gmap.gff | gffread -g IR64.fasta -w pla1_IR64_gmap_mRNA.fasta
----

. Run the multiple sequence aligment and SNP extraction with the full transcript sequences.
+
[source,sh]
----
cat pla1_Nipponbare_gmap_mRNA.fasta pla1_IR64_gmap_mRNA.fasta | mafft - > pla1_gmap_mafft.fasta
snp-sites -v pla1_gmap_mafft.fasta
----

== Genebank genomics portal

. Visit https://bridge.ipk-gatersleben.de/#geomap[BRIDGE], the barley genebank genomics portal.

== Analysis of RNA-seq of a wheat mutant and its wildtype 

=== Creating a kallisto index

. Create the project directory and change to it.
+
[source,sh]
----
mkdir ~/wheat_rnaseq
cd ~/wheat_rnaseq
----

. Create symbolic links to the annotation files and decompress them.
+
[source,sh]
----
ln -s /home/iamz/materials_Mascher/iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa.zip 
ln -s /home/iamz/materials_Mascher/Scripting/iwgsc_refseqv1.0_FunctionalAnnotation_v1.zip 
----

. Decompress the zip archives.
+
[source,sh]
----
unzip iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa.zip 
unzip iwgsc_refseqv1.0_FunctionalAnnotation_v1.zip 
----

. Count the number of sequences in the file. 
+
[source,sh]
----
grep -c '^>' iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa
----

. Create an index for alignment with https://pachterlab.github.io/kallisto/[kallisto].
+
[source,sh]
----
kallisto index --index wheat_index iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa > kallisto_index.out 2>  kallisto_index.err & 
----

=== Quantifying transcript abundance

. Create symbolic links to read files.
+
[source,sh]
----
ln -s /home/iamz/materials_Mascher/GA*gz .
ln -s /home/iamz/materials_Mascher/WA*gz .
----

. Run the quantification for a single sample.
+
[source,sh]
----
kallisto quant --index wheat_index GA_0908-N_1_R1.fastq.gz GA_0908-N_1_R2.fastq.gz --output GA_0908-N_1_kallisto > GA_0908-N_1_kallisto.out 2> GA_0908-N_1_kallisto.err & 
----

. Run the quantification for all samples using a loop. This will take about three hours.
+
[source,sh]
----
 find | grep R1 | cut -d _ -f -3 | sort | while read i; do
  kallisto quant --index wheat_index  ${i}_R1.fastq.gz ${i}_R2.fastq.gz --output ${i}_kallisto > ${i}_kallisto.out 2>  ${i}_kallisto.err 
 done
----

. Check that there are results for all samples.
+
[source,sh]
----
head GA_0908-N_1_kallisto/abundance.tsv | column -t 
find | grep abundance.tsv | wc  -l 
find | grep abundance.tsv | xargs wc -l
grep -c '^>' iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa
----

. If there are issues, download the pre-computed kallisto results. Delete the current output directories. Unzip the archive.
+
[source,sh]
----
wget 'https://files.ipk-gatersleben.de/file/RBWEFkcdORboj1QM/zrsvWx9hzVxeEWN2/kallisto.zip'
rm -rf *_kallisto
unzip kallisto.zip
----

. Further analyses will be run https://github.com/wyguo/ThreeDRNAseq[3D RNA-seq] (https://www.biorxiv.org/content/10.1101/656686v1[paper]). 
We will now create the necessary input file.

. Create tables with the metadata.
+
[source,sh]
----
grep '^>' iwgsc_refseqv1.0_HighConf_CDS_2017Mar13.fa | tr -d '>' | cut -d ' ' -f 1 > transcripts.txt 
cut -d . -f 1 transcripts.txt > genes.txt 
paste -d , transcripts.txt genes.txt  > transcript_genes.csv #<1>

find -type d | grep kallisto  | cut -d / -f 2 | sort > kallisto.txt #<2>

cat kallisto.txt  | tr _- '\t' | awk '{print $1","$3","$4}' | paste -d , - kallisto.txt | awk 'BEGIN{print "stage,allele,rep,folder"} {print}' > sample_info.csv #<3>
----
<1> Assignment of transcript isoforms to genes.
<2> List of Kallisto output directories.
<3> Table with experimental design.

. Create a zip archiv containing the kallisto output folders. Not needed if you use the download "kallisto.zip" file.
+
[source,sh]
----
zip -r kallisto.zip *_kallisto #<1>
----
<1> `-r` instruct `zip` to travel the directory structure recursively, that is, to include all files in the folders.

//. Create a new folders on your Windows desktop and copy the files there.
//+
//[source,sh]
//----
//mkdir /media/Desktop/3drna
//cp -t /media/Desktop/3drna sample_info.csv transcript_genes.csv kallisto.zip
//----

. Open the https://3drnaseq.hutton.ac.uk/app_direct/3DRNAseq/[3D RNA-seq app].

. Upload the datasets in the *Data generation* tab.

. Follow the 3D DNA-seq steps. Click on the question mark symbol to get more guidance.

== Transcriptome atlases

. Visit http://bar.utoronto.ca[BAR], the Bio-Analytic Resource for Plant Biology (https://link.springer.com/protocol/10.1007%2F978-1-4939-6658-5_6[paper]).

== Barley epigenome browser

. Visit the https://ics.hutton.ac.uk/barley-epigenome/[Barley epigenome browser].

== Cereal pangenomes at grain genes

. Cereal pangenomes are hosted at https://wheat.pw.usda.gov/GG3/[GrainGenes].
